File: dart/lib/latlngz.dart
/// Coordinate triple
class LatLngZ {
  final double lat;
  final double lng;
  final double z;

  LatLngZ(this.lat, this.lng, [this.z = 0]);

  @override
  String toString() => "LatLngZ [lat=$lat, lng=$lng, z=$z]";

  @override
  bool operator ==(other) =>
      other is LatLngZ && other.lat == lat && other.lng == lng && other.z == z;

  @override
  int get hashCode => lat.hashCode + lng.hashCode + z.hashCode;
}

/// 3rd dimension specification.
/// Example a level, altitude, elevation or some other custom value.
/// ABSENT is default when there is no third dimension en/decoding required.
enum ThirdDimension {
  ABSENT,
  LEVEL,
  ALTITUDE,
  ELEVATION,
  RESERVED1,
  RESERVED2,
  CUSTOM1,
  CUSTOM2
}

File: dotnet/src/LatLngZ.cs
using System;

namespace HERE.FlexiblePolyline
{
    /// <summary>
    /// Coordinate triple
    /// </summary>
    public class LatLngZ
    {
        public double Lat { get; }
        public double Lng { get; }
        public double Z { get; }

        public LatLngZ(double latitude, double longitude, double thirdDimension = 0)
        {
            Lat = latitude;
            Lng = longitude;
            Z = thirdDimension;
        }

        public override string ToString()
        {
            return "LatLngZ [lat=" + Lat + ", lng=" + Lng + ", z=" + Z + "]";
        }

        public override bool Equals(object obj)
        {
            if (this == obj)
            {
                return true;
            }

            if (obj is LatLngZ latLngZ)
            {
                if (latLngZ.Lat == Lat && latLngZ.Lng == Lng && latLngZ.Z == Z)
                {
                    return true;
                }
            }

            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Lat, Lng, Z);
        }
    }
}
